# KEDRO SETUP + IMPORT

In [56]:
%load_ext kedro.ipython

The kedro.ipython extension is already loaded. To reload it, use:
  %reload_ext kedro.ipython


In [57]:
%reload_kedro

[07/24/24 15:06:20] INFO     Resolved project path as:                                              ]8;id=802021;file://c:\Users\PC\.conda\envs\goodreadv2\lib\site-packages\kedro\ipython\__init__.py\__init__.py]8;;\:]8;id=150577;file://c:\Users\PC\.conda\envs\goodreadv2\lib\site-packages\kedro\ipython\__init__.py#171\171]8;;\
                             C:\Users\PC\Desktop\PRO-G\ECOLE\DSTI_goodreads\goodreads-predictor.                   
                             To set a different path, run '%reload_kedro <project_root>'                           

[07/24/24 15:06:21] INFO     Kedro project goodreads-predictor                                      ]8;id=176464;file://c:\Users\PC\.conda\envs\goodreadv2\lib\site-packages\kedro\ipython\__init__.py\__init__.py]8;;\:]8;id=409522;file://c:\Users\PC\.conda\envs\goodreadv2\lib\site-packages\kedro\ipython\__init__.py#141\141]8;;\

                    INFO     Defined global variable 'context', 'session', 'catalog' and            ]8;id=635420;file://c:\Users\PC\.conda\envs\goodreadv2\lib\site-packages\kedro\ipython\__init__.py\__init__.py]8;;\:]8;id=290790;file://c:\Users\PC\.conda\envs\goodreadv2\lib\site-packages\kedro\ipython\__init__.py#142\142]8;;\
                             'pipelines'                                                                           

                    INFO     Registered line magic 'run_viz'                                        ]8;id=658480;file://c:\Users\PC\.conda\envs\goodreadv2\lib\site-packages\kedro\ipython\__init__.py\__init__.py]8;;\:]8;id=105186;file://c:\Users\PC\.conda\envs\goodreadv2\lib\site-packages\kedro\ipython\__init__.py#148\148]8;;\

In [58]:
import pandas as pd
import re
import numpy as np
import matplotlib.pyplot as plt
import plotly.graph_objects as go # For the waterfall plot -- cannot easily do this in matplotlib
from typing import List, Dict, Tuple, Any



In [59]:
context.catalog.list()


[
    'books_raw',
    'books_loaded',
    'publisher_consolidation',
    'price_by_isbn_input',
    'price_by_isbn',
    'title_embeddings_original',
    'authors_embeddings',
    'description_embeddings',
    'open_library_book_api_info_input',
    'open_library_book_api_info',
    'filtered_books',
    'exclusions_summary',
    'scope_waterfall_plot',
    'parameters',
    'params:title_column',
    'params:author_column',
    'params:description_column',
    'params:original_data_key_column',
    'params:color_gr_brown',
    'params:color_gr_purple',
    'params:color_gr_green',
    'params:color_gr_tan_background'
]

# DATA IMPORT

In [60]:
books_input = catalog.load('books_loaded')
price_by_isbn = catalog.load('price_by_isbn')
title_embeddings = catalog.load('title_embeddings_original')
book_api_data = catalog.load('open_library_book_api_info')
authors_embeddings = catalog.load('authors_embeddings')

# Get the color values
color_gr_brown = catalog.load("params:color_gr_brown")
color_gr_purple = catalog.load("params:color_gr_purple")
color_gr_green = catalog.load("params:color_gr_green")
color_gr_tan_background = catalog.load("params:color_gr_tan_background")

# Add in the price data
books_input = books_input.merge(price_by_isbn, on='isbn13', how='left')

                    INFO     Loading data from books_loaded (ParquetDataset)...                 ]8;id=62589;file://c:\Users\PC\.conda\envs\goodreadv2\lib\site-packages\kedro\io\data_catalog.py\data_catalog.py]8;;\:]8;id=941039;file://c:\Users\PC\.conda\envs\goodreadv2\lib\site-packages\kedro\io\data_catalog.py#508\508]8;;\

                    INFO     Loading data from price_by_isbn (ParquetDataset)...                ]8;id=333859;file://c:\Users\PC\.conda\envs\goodreadv2\lib\site-packages\kedro\io\data_catalog.py\data_catalog.py]8;;\:]8;id=861246;file://c:\Users\PC\.conda\envs\goodreadv2\lib\site-packages\kedro\io\data_catalog.py#508\508]8;;\

                    INFO     Loading data from title_embeddings_original (ParquetDataset)...    ]8;id=869080;file://c:\Users\PC\.conda\envs\goodreadv2\lib\site-packages\kedro\io\data_catalog.py\data_catalog.py]8;;\:]8;id=668944;file://c:\Users\PC\.conda\envs\goodreadv2\lib\site-packages\kedro\io\data_catalog.py#508\508]8;;\

                    INFO     Loading data from open_library_book_api_info (PickleDataset)...    ]8;id=450612;file://c:\Users\PC\.conda\envs\goodreadv2\lib\site-packages\kedro\io\data_catalog.py\data_catalog.py]8;;\:]8;id=789172;file://c:\Users\PC\.conda\envs\goodreadv2\lib\site-packages\kedro\io\data_catalog.py#508\508]8;;\

[07/24/24 15:06:22] INFO     Loading data from authors_embeddings (ParquetDataset)...           ]8;id=412360;file://c:\Users\PC\.conda\envs\goodreadv2\lib\site-packages\kedro\io\data_catalog.py\data_catalog.py]8;;\:]8;id=60819;file://c:\Users\PC\.conda\envs\goodreadv2\lib\site-packages\kedro\io\data_catalog.py#508\508]8;;\

                    INFO     Loading data from params:color_gr_brown (MemoryDataset)...         ]8;id=167283;file://c:\Users\PC\.conda\envs\goodreadv2\lib\site-packages\kedro\io\data_catalog.py\data_catalog.py]8;;\:]8;id=761513;file://c:\Users\PC\.conda\envs\goodreadv2\lib\site-packages\kedro\io\data_catalog.py#508\508]8;;\

                    INFO     Loading data from params:color_gr_purple (MemoryDataset)...        ]8;id=153651;file://c:\Users\PC\.conda\envs\goodreadv2\lib\site-packages\kedro\io\data_catalog.py\data_catalog.py]8;;\:]8;id=91075;file://c:\Users\PC\.conda\envs\goodreadv2\lib\site-packages\kedro\io\data_catalog.py#508\508]8;;\

                    INFO     Loading data from params:color_gr_green (MemoryDataset)...         ]8;id=304856;file://c:\Users\PC\.conda\envs\goodreadv2\lib\site-packages\kedro\io\data_catalog.py\data_catalog.py]8;;\:]8;id=553282;file://c:\Users\PC\.conda\envs\goodreadv2\lib\site-packages\kedro\io\data_catalog.py#508\508]8;;\

                    INFO     Loading data from params:color_gr_tan_background                   ]8;id=474039;file://c:\Users\PC\.conda\envs\goodreadv2\lib\site-packages\kedro\io\data_catalog.py\data_catalog.py]8;;\:]8;id=166131;file://c:\Users\PC\.conda\envs\goodreadv2\lib\site-packages\kedro\io\data_catalog.py#508\508]8;;\
                             (MemoryDataset)...                                                                    

In [61]:
data = books_input

In [62]:
df = data.merge(title_embeddings)


In [63]:
df = data.merge(authors_embeddings)

# ANALYSIS & VIZUALISATION

## *INTRO*

In [64]:
df.head(5)

bookID                                              title  \
0       1  Harry Potter and the Half-Blood Prince (Harry ...   
1       2  Harry Potter and the Order of the Phoenix (Har...   
2       4  Harry Potter and the Chamber of Secrets (Harry...   
3       5  Harry Potter and the Prisoner of Azkaban (Harr...   
4       8  Harry Potter Boxed Set  Books 1-5 (Harry Potte...   

                      authors  average_rating        isbn         isbn13  \
0  J.K. Rowling/Mary GrandPré            4.57  0439785960  9780439785969   
1  J.K. Rowling/Mary GrandPré            4.49  0439358078  9780439358071   
2                J.K. Rowling            4.42  0439554896  9780439554893   
3  J.K. Rowling/Mary GrandPré            4.56  043965548X  9780439655484   
4  J.K. Rowling/Mary GrandPré            4.78  0439682584  9780439682589   

  language_code    num_pages  ratings_count  text_reviews_count  \
0           eng          652        2095690               27591   
1           eng          870        2153167               29221   
2           eng          352           6333                 244   
3           eng          435        2339585               36325   
4           eng         2690          41428                 164   

  publication_date        publisher  Price PriceStrategy  \
0        9/16/2006  Scholastic Inc.  12.99    LIST_PRICE   
1         9/1/2004  Scholastic Inc.  99.99    LIST_PRICE   
2        11/1/2003       Scholastic  24.95    LIST_PRICE   
3         5/1/2004  Scholastic Inc.   7.99    LIST_PRICE   
4        9/13/2004       Scholastic  40.95    LIST_PRICE   

                                   authors_embedding  
0  [-0.30315670371055603, 0.11502666771411896, -0...  
1  [-0.30315670371055603, 0.11502666771411896, -0...  
2  [-0.22978533804416656, 0.1457509994506836, -0....  
3  [-0.30315670371055603, 0.11502666771411896, -0...  
4  [-0.30315670371055603, 0.11502666771411896, -0...

In [65]:
df['authors'].nunique()

6639

In [66]:
first_hypodf=df[['authors', 'average_rating', 'authors_embedding']]

In [67]:
first_hypodf.describe()

average_rating
count    11123.000000
mean         3.934075
std          0.350485
min          0.000000
25%          3.770000
50%          3.960000
75%          4.140000
max          5.000000

### FIRST THING NOTICED 

THE DATA IS BIASED BY TE QUALITY OF BOOKS, HAVING A MEAN OF 3.8 OUT OF 5 MEANS THAT WE DON'T HAVE ENOUGH EXAMPLES OF NOT SO GOOD BOOKS.

## *CATCHING SEMANTIC RELATIONS : DO EMBEDDINGS WORK ?*

In [68]:
from sklearn.metrics.pairwise import cosine_similarity
"""
Trying to see if the embedding catches similarities between different instances of a same author
"""

one = first_hypodf['authors_embedding'].iloc[0]
two = first_hypodf['authors_embedding'].iloc[2]

one = one.reshape(1,-1)
two = two.reshape(1,-1)



In [71]:
cosine_similarity(one,two)

array([[0.75045805]])

A cosine similarity of 0.75 is very good for two instances fairly different. We can use this technique to merge names together. For example we can set a threshold of 70% and basically say that if the cosine similarity of two embeddings is greater or equal to 70 the author is the same or has some differences (a collaborator like the Mary GrandPré which is the illustrator of the latests HP books.)

## *MERGING AUTHORS BASED ON THEIR EMBEDDINGS*

In [72]:
# Basically converting all authors embedding to match a cosine computation

first_hypodf.loc[:, 'authors_embedding'] = first_hypodf['authors_embedding'].apply(lambda x: x.reshape(1, -1))

In [75]:
def standardize_author_names(df, threshold=0.72):
    embeddings = first_hypodf['authors_embedding'].tolist()
    authors = first_hypodf['authors'].tolist()
    
    
    embeddings = [embedding.flatten() for embedding in embeddings]
    
    for i in range(len(embeddings)):
        current_embedding = embeddings[i].reshape(1, -1)
        similarities = cosine_similarity(current_embedding, embeddings)
        
        for j in range(len(similarities[0])):
            if i != j and similarities[0][j] >= threshold:
                authors[i] = authors[j]
                break
    
    first_hypodf.loc[:, 'authors'] = authors

In [76]:
standardize_author_names(first_hypodf)

In [77]:
first_hypodf['authors'].nunique()

1567

We went from 6639 authors name to 1567 in a robust manner this is actually very good to analyse with ease 

# FHDF, STARTING TO WORK WITH THE DATA FOR MODELING

In [123]:
fhdf=first_hypodf

In [124]:
fhdf

authors  average_rating  \
0                            J.K. Rowling/Mary GrandPré            4.57   
1                            J.K. Rowling/Mary GrandPré            4.49   
2                            J.K. Rowling/Mary GrandPré            4.42   
3                            J.K. Rowling/Mary GrandPré            4.56   
4                            J.K. Rowling/Mary GrandPré            4.78   
...                                                 ...             ...   
11118  John Milton/Merritt Y. Hughes/David Scott Kastan            4.06   
11119        Henry James/Gabriel Brownstein/Mary Cregan            4.08   
11120        Henry James/Gabriel Brownstein/Mary Cregan            3.96   
11121        Henry James/Gabriel Brownstein/Mary Cregan            3.72   
11122                       Mark Twain/Michael Prichard            3.91   

                                       authors_embedding  \
0      [[-0.30315670371055603, 0.11502666771411896, -...   
1      [[-0.30315670371055603, 0.11502666771411896, -...   
2      [[-0.22978533804416656, 0.1457509994506836, -0...   
3      [[-0.30315670371055603, 0.11502666771411896, -...   
4      [[-0.30315670371055603, 0.11502666771411896, -...   
...                                                  ...   
11118  [[-0.403737336397171, 0.2087896317243576, -0.1...   
11119  [[-0.5178908705711365, 0.5561336278915405, -0....   
11120  [[-0.5178909301757812, 0.5561334490776062, -0....   
11121  [[-0.5178908705711365, 0.5561336278915405, -0....   
11122  [[0.33673834800720215, 0.24192380905151367, 0....   

                                                   title         isbn13  
0      Harry Potter and the Half-Blood Prince (Harry ...  9780439785969  
1      Harry Potter and the Order of the Phoenix (Har...  9780439358071  
2      Harry Potter and the Chamber of Secrets (Harry...  9780439554893  
3      Harry Potter and the Prisoner of Azkaban (Harr...  9780439655484  
4      Harry Potter Boxed Set  Books 1-5 (Harry Potte...  9780439682589  
...                                                  ...            ...  
11118   Expelled from Eden: A William T. Vollmann Reader  9781560254416  
11119                        You Bright and Risen Angels  9780140110876  
11120                    The Ice-Shirt (Seven Dreams #1)  9780140131963  
11121                                        Poor People  9780060878825  
11122                        Las aventuras de Tom Sawyer  9788497646987  

[11123 rows x 5 columns]

In [125]:
fhdf.loc[:,'title']=df['title']

In [126]:
title_embeddings

isbn13                                    title_embedding
0      9780439785969  [-0.15156979858875275, 0.4935136139392853, 0.0...
1      9780439358071  [-0.22418256103992462, 0.20987719297409058, 0....
2      9780439554893  [-0.054178591817617416, 0.22646689414978027, -...
3      9780439655484  [-0.2782197892665863, 0.261374831199646, -0.12...
4      9780439682589  [-0.16637001931667328, 0.2694065272808075, -0....
...              ...                                                ...
11118  9781560254416  [-0.13434666395187378, 0.08153267949819565, 0....
11119  9780140110876  [0.32915258407592773, 0.15614797174930573, 0.0...
11120  9780140131963  [-0.02630615048110485, 0.1311979591846466, -0....
11121  9780060878825  [-0.10792334377765656, 0.27181512117385864, -0...
11122  9788497646987  [0.0736294761300087, -0.2517438232898712, -0.2...

[11123 rows x 2 columns]

In [127]:
fhdf.loc[:,'isbn13'] = df['isbn13']
fhdf = fhdf.merge(title_embeddings, on='isbn13', how='left')
fhdf.loc[:,'publisher'] = df['publisher']
fhdf.loc[:,'price'] = df['Price']

In [130]:
fhdf

authors  average_rating  \
0                            J.K. Rowling/Mary GrandPré            4.57   
1                            J.K. Rowling/Mary GrandPré            4.49   
2                            J.K. Rowling/Mary GrandPré            4.42   
3                            J.K. Rowling/Mary GrandPré            4.56   
4                            J.K. Rowling/Mary GrandPré            4.78   
...                                                 ...             ...   
11118  John Milton/Merritt Y. Hughes/David Scott Kastan            4.06   
11119        Henry James/Gabriel Brownstein/Mary Cregan            4.08   
11120        Henry James/Gabriel Brownstein/Mary Cregan            3.96   
11121        Henry James/Gabriel Brownstein/Mary Cregan            3.72   
11122                       Mark Twain/Michael Prichard            3.91   

                                       authors_embedding  \
0      [[-0.30315670371055603, 0.11502666771411896, -...   
1      [[-0.30315670371055603, 0.11502666771411896, -...   
2      [[-0.22978533804416656, 0.1457509994506836, -0...   
3      [[-0.30315670371055603, 0.11502666771411896, -...   
4      [[-0.30315670371055603, 0.11502666771411896, -...   
...                                                  ...   
11118  [[-0.403737336397171, 0.2087896317243576, -0.1...   
11119  [[-0.5178908705711365, 0.5561336278915405, -0....   
11120  [[-0.5178909301757812, 0.5561334490776062, -0....   
11121  [[-0.5178908705711365, 0.5561336278915405, -0....   
11122  [[0.33673834800720215, 0.24192380905151367, 0....   

                                                   title         isbn13  \
0      Harry Potter and the Half-Blood Prince (Harry ...  9780439785969   
1      Harry Potter and the Order of the Phoenix (Har...  9780439358071   
2      Harry Potter and the Chamber of Secrets (Harry...  9780439554893   
3      Harry Potter and the Prisoner of Azkaban (Harr...  9780439655484   
4      Harry Potter Boxed Set  Books 1-5 (Harry Potte...  9780439682589   
...                                                  ...            ...   
11118   Expelled from Eden: A William T. Vollmann Reader  9781560254416   
11119                        You Bright and Risen Angels  9780140110876   
11120                    The Ice-Shirt (Seven Dreams #1)  9780140131963   
11121                                        Poor People  9780060878825   
11122                        Las aventuras de Tom Sawyer  9788497646987   

                                         title_embedding        publisher  \
0      [-0.15156979858875275, 0.4935136139392853, 0.0...  Scholastic Inc.   
1      [-0.22418256103992462, 0.20987719297409058, 0....  Scholastic Inc.   
2      [-0.054178591817617416, 0.22646689414978027, -...       Scholastic   
3      [-0.2782197892665863, 0.261374831199646, -0.12...  Scholastic Inc.   
4      [-0.16637001931667328, 0.2694065272808075, -0....       Scholastic   
...                                                  ...              ...   
11118  [-0.13434666395187378, 0.08153267949819565, 0....    Da Capo Press   
11119  [0.32915258407592773, 0.15614797174930573, 0.0...    Penguin Books   
11120  [-0.02630615048110485, 0.1311979591846466, -0....    Penguin Books   
11121  [-0.10792334377765656, 0.27181512117385864, -0...             Ecco   
11122  [0.0736294761300087, -0.2517438232898712, -0.2...    Edimat Libros   

        price  
0      12.990  
1      99.990  
2      24.950  
3       7.990  
4      40.950  
...       ...  
11118  22.990  
11119  25.000  
11120  24.000  
11121  29.950  
11122   5.305  

[11123 rows x 8 columns]

# CORR COEF BETWEEN FEATURES

In [133]:
"""
Flattening the embedding to be able to work with them from now on!
"""
fhdf['authors_embedding'] = [np.array(embedding).flatten() for embedding in fhdf['authors_embedding']]

In [134]:
fhdf

authors  average_rating  \
0                            J.K. Rowling/Mary GrandPré            4.57   
1                            J.K. Rowling/Mary GrandPré            4.49   
2                            J.K. Rowling/Mary GrandPré            4.42   
3                            J.K. Rowling/Mary GrandPré            4.56   
4                            J.K. Rowling/Mary GrandPré            4.78   
...                                                 ...             ...   
11118  John Milton/Merritt Y. Hughes/David Scott Kastan            4.06   
11119        Henry James/Gabriel Brownstein/Mary Cregan            4.08   
11120        Henry James/Gabriel Brownstein/Mary Cregan            3.96   
11121        Henry James/Gabriel Brownstein/Mary Cregan            3.72   
11122                       Mark Twain/Michael Prichard            3.91   

                                       authors_embedding  \
0      [-0.30315670371055603, 0.11502666771411896, -0...   
1      [-0.30315670371055603, 0.11502666771411896, -0...   
2      [-0.22978533804416656, 0.1457509994506836, -0....   
3      [-0.30315670371055603, 0.11502666771411896, -0...   
4      [-0.30315670371055603, 0.11502666771411896, -0...   
...                                                  ...   
11118  [-0.403737336397171, 0.2087896317243576, -0.12...   
11119  [-0.5178908705711365, 0.5561336278915405, -0.0...   
11120  [-0.5178909301757812, 0.5561334490776062, -0.0...   
11121  [-0.5178908705711365, 0.5561336278915405, -0.0...   
11122  [0.33673834800720215, 0.24192380905151367, 0.0...   

                                                   title         isbn13  \
0      Harry Potter and the Half-Blood Prince (Harry ...  9780439785969   
1      Harry Potter and the Order of the Phoenix (Har...  9780439358071   
2      Harry Potter and the Chamber of Secrets (Harry...  9780439554893   
3      Harry Potter and the Prisoner of Azkaban (Harr...  9780439655484   
4      Harry Potter Boxed Set  Books 1-5 (Harry Potte...  9780439682589   
...                                                  ...            ...   
11118   Expelled from Eden: A William T. Vollmann Reader  9781560254416   
11119                        You Bright and Risen Angels  9780140110876   
11120                    The Ice-Shirt (Seven Dreams #1)  9780140131963   
11121                                        Poor People  9780060878825   
11122                        Las aventuras de Tom Sawyer  9788497646987   

                                         title_embedding        publisher  \
0      [-0.15156979858875275, 0.4935136139392853, 0.0...  Scholastic Inc.   
1      [-0.22418256103992462, 0.20987719297409058, 0....  Scholastic Inc.   
2      [-0.054178591817617416, 0.22646689414978027, -...       Scholastic   
3      [-0.2782197892665863, 0.261374831199646, -0.12...  Scholastic Inc.   
4      [-0.16637001931667328, 0.2694065272808075, -0....       Scholastic   
...                                                  ...              ...   
11118  [-0.13434666395187378, 0.08153267949819565, 0....    Da Capo Press   
11119  [0.32915258407592773, 0.15614797174930573, 0.0...    Penguin Books   
11120  [-0.02630615048110485, 0.1311979591846466, -0....    Penguin Books   
11121  [-0.10792334377765656, 0.27181512117385864, -0...             Ecco   
11122  [0.0736294761300087, -0.2517438232898712, -0.2...    Edimat Libros   

        price  
0      12.990  
1      99.990  
2      24.950  
3       7.990  
4      40.950  
...       ...  
11118  22.990  
11119  25.000  
11120  24.000  
11121  29.950  
11122   5.305  

[11123 rows x 8 columns]

-> To correlate things together we should turn the embedding into a scalar

In [135]:
fhdf['mean_author_embedding'] = fhdf['authors_embedding'].apply(np.mean)
fhdf['mean_title_embedding'] = fhdf['title_embedding'].apply(np.mean)

In [136]:
fhdf['mean_author_embedding'].corr(fhdf['average_rating'])

-0.05092749661907622

# DATA PREP FOR MODELING

In [287]:
encoded_df = fhdf[["mean_author_embedding","authors_embedding", "price", 'publisher']]

In [288]:
encoded_df

mean_author_embedding  \
0                  -0.001366   
1                  -0.001366   
2                  -0.000989   
3                  -0.001366   
4                  -0.001366   
...                      ...   
11118               0.000349   
11119              -0.000150   
11120              -0.000150   
11121              -0.000150   
11122              -0.001303   

                                       authors_embedding   price  \
0      [-0.30315670371055603, 0.11502666771411896, -0...  12.990   
1      [-0.30315670371055603, 0.11502666771411896, -0...  99.990   
2      [-0.22978533804416656, 0.1457509994506836, -0....  24.950   
3      [-0.30315670371055603, 0.11502666771411896, -0...   7.990   
4      [-0.30315670371055603, 0.11502666771411896, -0...  40.950   
...                                                  ...     ...   
11118  [-0.403737336397171, 0.2087896317243576, -0.12...  22.990   
11119  [-0.5178908705711365, 0.5561336278915405, -0.0...  25.000   
11120  [-0.5178909301757812, 0.5561334490776062, -0.0...  24.000   
11121  [-0.5178908705711365, 0.5561336278915405, -0.0...  29.950   
11122  [0.33673834800720215, 0.24192380905151367, 0.0...   5.305   

             publisher  
0      Scholastic Inc.  
1      Scholastic Inc.  
2           Scholastic  
3      Scholastic Inc.  
4           Scholastic  
...                ...  
11118    Da Capo Press  
11119    Penguin Books  
11120    Penguin Books  
11121             Ecco  
11122    Edimat Libros  

[11123 rows x 4 columns]

In [289]:

#embeddings = np.array(encoded_df['authors_embedding'].tolist())
#embeddings_df = pd.DataFrame(embeddings, columns=[f'emb_{i}' for i in range(embeddings.shape[1])])

# Encodage One-Hot de la colonne 'publisher'
publisher_df = pd.get_dummies(encoded_df, columns=['publisher'])
publisher_df = publisher_df.drop(['authors_embedding', 'price', 'mean_author_embedding'], axis=1)

# Combiner toutes les features
features_df = pd.concat([encoded_df['mean_author_embedding'], encoded_df[['price']], publisher_df], axis=1)


X = features_df
y = fhdf['average_rating']


In [290]:
X = features_df
y = fhdf['average_rating']

In [291]:
print("Shape of X:", X.shape)
print("Length of y:", len(y))


Shape of X: (11123, 2292)
Length of y: 11123


In [292]:
# Convertir X et y en DataFrames
X_df = pd.DataFrame(X)
y_df = pd.DataFrame(y, columns=['average_rating'])

# Identifier les lignes contenant des NaN dans X
nan_indices = X_df[X_df.isna().any(axis=1)].index

# Supprimer les lignes correspondantes de X et y
X_df_cleaned = X_df.drop(nan_indices)
y_df_cleaned = y_df.drop(nan_indices)

# Assurez-vous que les indices sont maintenant alignés
X_cleaned = X_df_cleaned.values
y_cleaned = y_df_cleaned.values.ravel()

# Vérifier les formes après nettoyage
print("Shape of X_cleaned:", X_cleaned.shape)
print("Shape of y_cleaned:", y_cleaned.shape)

Shape of X_cleaned: (11112, 2292)
Shape of y_cleaned: (11112,)


# MODELING

In [278]:
import sklearn
from sklearn.model_selection import train_test_split

In [293]:
X_train, X_test, y_train, y_test = train_test_split(X_cleaned , y_cleaned , test_size=0.25)

In [294]:
from sklearn.neighbors import KNeighborsRegressor

In [295]:
reg = KNeighborsRegressor()

In [296]:
model = reg.fit(X_train, y_train)

In [297]:
model

KNeighborsRegressor()

In [298]:
y_pred = model.predict(X_test)

# RESULTS (METRICS & COMMENTARIES)

In [299]:
from sklearn.metrics import mean_squared_error, r2_score

mean_squared_error(y_test, y_pred) , r2_score(y_test, y_pred)

(1.1786912915166128, -0.09640319235479633)